nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+                i, "only views" if only_merge_ops_with_view else "all",
+                before, after, t_end - t_start)
+
+        self._reinitialize_model_ops(self.dg)
+
+    def _perform_single_pass(self, only_merge_ops_with_view):
+        """Perform a single optimization pass.
+
+        Parameters
+        ----------
+        only_merge_ops_with_view : bool
+            Limits operator merges to operators with views.
+        """
+        self._init_pass_helper_vars()
+        op_replacements, sig_replacements = self._perform_merges(
+            only_merge_ops_with_view)
+        ops = list(op_replacements.values())
+        self._replace_op_signals(ops, sig_replacements)
+        self._update_dg(op_replacements)
+        self._update_model_sigs(sig_replacements)
+
+    def _init_pass_helper_vars(self):
+        """Initializes the internal helper variables for an optimization pass.
+        """
+        self._sig2op = defaultdict(list)
+        for op in self.dg:
+            for s in op.all_signals:
+                self._sig2op[s].append(op)
+        self._merged = set()
+
+    def _ops_by_type(self, ops):
+        """Returns `ops` as a dictionary sorted by type."""
+        by_type = defaultdict(list)
+        for op in ops:
+            by_type[type(op)].append(op)
+        return by_type
+
+    def _log_counts(self):
+        """Logs for each operator type the count in the dependency graph."""
+        if logger.isEnabledFor(logging.DEBUG):
+            for tp, ops in self._ops_by_type(self.dg).items():
+                logger.debug("%s: %i", tp, len(ops))
+
+    def _perform_merges(self, only_merge_ops_with_view):
+        """Go through all operators and merge them where possible.
+
+        Parameters
+        ----------
+        only_merge_ops_with_view : bool
+            Limit merges to operators with views.
+
+        Returns
+        -------
+        op_replacements : dict
+            Dictionary that maps from original operators to the newly merged
+            operators. To be used to update the dependency graph.
+        sig_replacements : dict
+            Dictionary that maps from original signals to new signals resulting
+            from the merges. To be used to update the signals referenced by
+            the operators.
+        """
+        step_order = toposort(self.dg)
+        tc = transitive_closure(self.dg, step_order)
+        by_type = self._ops_by_type(step_order)
+
+        op_replacements = {}
+        sig_replacements = {}
+
+        # Note that the following loops will stop as soon as any operators
+        # have been merged, so that merges are performed on at most one type
+        # of operators. This function will be called again once the dependency
+        # graph and signals have been updated with those merges to continue
+        # merging other operator types.
+
+        # Heuristic order to reduce runtime and number of passes
+        for tp in [ElementwiseInc, SlicedCopy, DotInc, SimNeurons]:
```

Comment:
There are other types of operators that can be merged, but these are done in the next loop in random order.